Server won't unset the resources_assigned attribute during resource unset anymore, if present

Bhagat-Rajput opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Queue resources_assigned goes negative because we unset a resource which is already been occupied by the job.
**Steps to reproduce:**
   1.  set q <queue_name> resources_available.ncpus=<value>
   2.  run a job
   3. unset q <qeueu_name> resources_available.ncpus
   4. qdel the job
   5. qstat -Qf <queue_name> and see "resources_assigned" value

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
As per the new changes, If we unset the resource then will remove only resources_available.<resc_name> attribute  and leave resources_assigned.<resc_name> attribute as it is with its original value((could be zero as well) if present. Resources_assigned.<resc_name> will be there in the system forever, until the resource is deleted or server can also unset this attribute if didn't find any jobs related to that particular resource in the system at the time of restart.

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->

PTL logs:- [ptl_que_resc_usage.txt](https://github.com/PBSPro/pbspro/files/4386420/ptl_que_resc_usage.txt)



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -167,3 +173,36 @@ def test_resources_assigned_with_zero_val(self):
+        # create a resource
+        self.create_custom_resc()
+        # submit jobs
+        j_attr = {ATTR_queue: 'workq',
```
No real reason to request workq.  It is the default.

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -167,3 +173,36 @@ def test_resources_assigned_with_zero_val(self):
+        j1 = Job(TEST_USER, j_attr)
+        j1.set_sleep_time(30)
+        jid_1 = self.server.submit(j1)
+        j2 = Job(TEST_USER, j_attr)
```
No need to request TEST_USER, it is the default.  Just says attrs=j_attr

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -167,3 +173,36 @@ def test_resources_assigned_with_zero_val(self):
+        j1.set_sleep_time(30)
+        jid_1 = self.server.submit(j1)
+        j2 = Job(TEST_USER, j_attr)
+        j2.set_sleep_time(30)
```
You aren't waiting for the job to finish, no need to set the sleep time.

Bhagat-Rajput replied:
Actually, here we are waiting for jobs to end after "except".

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -167,3 +173,36 @@ def test_resources_assigned_with_zero_val(self):
+        try:
+            self.server.manager(MGR_CMD_DELETE, RSC, id='foo')
+        except PbsManagerError as e:
+            self.assertTrue("Resource busy on job" in e.msg[0])
```
Use assertIn()

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -167,3 +173,36 @@ def test_resources_assigned_with_zero_val(self):
+        self.assertNotIn('resources_assigned.foo',
+                         q_status[0].keys(), self.err_msg)
+        self.assertNotIn('resources_available.foo',
+                         q_status[0].keys(), self.err_msg)
```
No reason to say .keys(), just use the dictionary itself.

bhroam commented on test/tests/functional/pbs_que_resc_usage.py:
```diff
@@ -50,6 +50,19 @@ def setUp(self):
+        self.scheduler.add_resource('foo')
+        self.server.manager(MGR_CMD_SET, QUEUE, {
+                            'resources_available.foo': 6}, id='workq')
+        q_status = self.server.status(QUEUE, id='workq')
```
I'm not sure you really need to do this.  If the manager() failed, it would raise an exception.

Bhagat-Rajput commented:
Thanks @bhroam for the review.
Since this PR was merged by that time so i've addressed all your comments in a new PR https://github.com/PBSPro/pbspro/pull/1619. Please review it. Thanks.
